`hacking.md`: Add library hacking instructions and add `make check` (#7509)
NixOS/nix: doc/manual/src/contributing/hacking.md

To install it in `$(pwd)/outputs` and test it:

```console
[nix-shell]$ make check
[nix-shell]$ make install
[nix-shell]$ make installcheck -j $NIX_BUILD_CORES
[nix-shell]$ ./outputs/out/bin/nix --version
nix (Nix) 3.0
```

> To iterate on for example `libexpr`, prioritizing the most relevant checks:

Reviewer: What exactly do you mean by "iterate on"? What part of the example makes it "prioritise" the most relevant checks?

Author: The parts where the otherwise implied targets are mentioned: `libexpr-tests_RUN`, and because of the way the `make` invocations are ordered.

I wonder if merging some of the make invocations would improve parallelism without accidentally delaying the thing you want to test.

> To iterate on for example `libexpr`, prioritizing the most relevant checks:

Reviewer: ```suggestion
For example, to run specific checks first when developing `libexpr`:
```

Is that what you mean?

```console
[nix-shell]$ J="-j $NIX_BUILD_CORES"

> [nix-shell]$ make $J libexpr-tests_RUN && make $J && make check && make install && make installcheck $J

Author: ```suggestion
[nix-shell]$ make $J libexpr-tests_RUN && make $J && make check && make install && make tests/lang.sh.test installcheck $J
```

This may make lang.sh run sooner.

Reviewer 2: There are indeed a few variations here and I wouldn't worry much about which is best :). We can always add more information later.

> [nix-shell]$ make $J libexpr-tests_RUN && make $J && make check && make install && make installcheck $J

Author: ```suggestion
[nix-shell]$ make $J libexpr-tests_RUN check default && make install && make installcheck $J
```

Maybe this improves parallelism in the first part, while prioritizing the most relevant targets.
